Lib/ast.py

```diff
@@ -96,17 +96,26 @@ def _convert(node):
     return _convert(node_or_string)
 
 
-def dump(node, annotate_fields=True, include_attributes=False):
+def dump(node, annotate_fields=True, include_attributes=False, indent=None):
     """
     Return a formatted dump of the tree in node.  This is mainly useful for
     debugging purposes.  If annotate_fields is true (by default),
     the returned string will show the names and the values for fields.
     If annotate_fields is false, the result string will be more compact by
     omitting unambiguous field names.  Attributes such as line
     numbers and column offsets are not dumped by default.  If this is wanted,
-    include_attributes can be set to true.
+    include_attributes can be set to true.  If indent is a non-negative
+    integer or string, then the tree will be pretty-printed with that indent
+    level.
     """
-    def _format(node):
+    def _format(node, level=0):
+        if indent is not None:
+            level += 1
+            prefix = '\n' + indent * level
```

Comment:
Recently, `xml.etree.ElementTree` added similar indent function in b5d3ceea48c181b3e2c6c67424317afed606bd39 where it used a list to store generated strings as an optimization https://github.com/python/cpython/blob/daa82d019c52e95c3c57275307918078c1c0ac81/Lib/xml/etree/ElementTree.py#L1211-L1221. Could this optimization help here too?

I used similar function here with `dump` and `dump_cached` in this gist https://gist.github.com/tirkarthi/94f5dcd90757bee982980afe94cec82e. Below are some numbers

```
(py38-venv) ➜  cpython git:(master) ✗ python -V
Python 3.8.0b4
(py38-venv) ➜  cpython git:(master) ✗ python -m pyperf timeit -s 'import ast; from ast_compare import dump; obj = ast.parse(open("Lib/ast.py").read())' 'dump(obj, indent=4)' -o dump.json
.....................
Mean +- std dev: 11.3 ms +- 0.7 ms
(py38-venv) ➜  cpython git:(master) ✗ python -m pyperf timeit -s 'import ast; from ast_compare import dump_cached; obj = ast.parse(open("Lib/ast.py").read())' 'dump_cached(obj, indent=4)' -o dump_cached.json
.....................
Mean +- std dev: 10.6 ms +- 0.7 ms
(py38-venv) ➜  cpython git:(master) ✗ python -m pyperf compare_to dump.json dump_cached.json --table
+-----------+---------+-----------------------------+
| Benchmark | dump    | dump_cached                 |
+===========+=========+=============================+
| timeit    | 11.3 ms | 10.6 ms: 1.07x faster (-6%) |
+-----------+---------+-----------------------------+
```

```
(py38-venv) ➜  cpython git:(master) ✗ python -m pyperf timeit -s 'import ast; from ast_compare import dump; obj = ast.parse("1 + 1 == 2" * 100)' 'dump(obj, indent=4)' -o dump_1_1.json
.....................
Mean +- std dev: 1.77 ms +- 0.10 ms
(py38-venv) ➜  cpython git:(master) ✗ python -m pyperf timeit -s 'import ast; from ast_compare import dump_cached; obj = ast.parse("1 + 1 == 2" * 100)' 'dump_cached(obj, indent=4)' -o dump_1_1_cached.json
.....................
Mean +- std dev: 1.63 ms +- 0.13 ms
(py38-venv) ➜  cpython git:(master) ✗ python -m pyperf compare_to dump_1_1.json dump_1_1_cached.json --table
+-----------+----------+-----------------------------+
| Benchmark | dump_1_1 | dump_1_1_cached             |
+===========+==========+=============================+
| timeit    | 1.77 ms  | 1.63 ms: 1.09x faster (-8%) |
+-----------+----------+-----------------------------+
```

Reply:
No, I do not think it is worth. This function is used mainly for debugging.